x/stakeibc/keeper/msg_server_create_trade_route_test.go

```diff
@@ -0,0 +1,231 @@
+
+	_, err := s.GetMsgServer().CreateTradeRoute(sdk.WrapSDKContext(s.Ctx), &msg)
+	s.Require().ErrorContains(err, "invalid authority")
+}
+
+// Tests creating a duplicate trade route
+func (s *KeeperTestSuite) TestCreateTradeRoute_Failure_DuplicateTradeRoute() {
+	msg, _ := s.SetupTestCreateTradeRoute()
+
+	// Store down a trade route so the tx hits a duplicate trade route error
+	s.App.StakeibcKeeper.SetTradeRoute(s.Ctx, types.TradeRoute{
+		RewardDenomOnRewardZone: RewardDenom,
+		HostDenomOnHostZone:     HostDenom,
+	})
+
+	_, err := s.GetMsgServer().CreateTradeRoute(sdk.WrapSDKContext(s.Ctx), &msg)
+	s.Require().ErrorContains(err, "Trade route already exists")
+}
+
+// Tests creating a trade route when the host zone or withdrawal address does not exist
+func (s *KeeperTestSuite) TestCreateTradeRoute_Failure_HostZoneNotRegistered() {
+	msg, _ := s.SetupTestCreateTradeRoute()
+
+	// Remove the host zone withdrawal address and confirm it fails
+	invalidHostZone := s.MustGetHostZone(HostChainId)
+	invalidHostZone.WithdrawalIcaAddress = ""
+	s.App.StakeibcKeeper.SetHostZone(s.Ctx, invalidHostZone)
+
+	_, err := s.GetMsgServer().CreateTradeRoute(sdk.WrapSDKContext(s.Ctx), &msg)
+	s.Require().ErrorContains(err, "withdrawal account not initialized on host zone")
+
+	// Remove the host zone completely and check that that also fails
+	s.App.StakeibcKeeper.RemoveHostZone(s.Ctx, HostChainId)
+
+	_, err = s.GetMsgServer().CreateTradeRoute(sdk.WrapSDKContext(s.Ctx), &msg)
+	s.Require().ErrorContains(err, "host zone not found")
+}
+
+// Tests creating a trade route where the ICA channels cannot be created
+// because the ICA connections do not exist
+func (s *KeeperTestSuite) TestCreateTradeRoute_Failure_ConnectionNotFound() {
+	// Test with non-existent reward connection
+	msg, _ := s.SetupTestCreateTradeRoute()
+	msg.StrideToRewardConnectionId = "connection-X"
+
+	// Remove the host zone completely and check that that also fails
+	_, err := s.GetMsgServer().CreateTradeRoute(sdk.WrapSDKContext(s.Ctx), &msg)
+	s.Require().ErrorContains(err, "unable to register the unwind ICA account: connection connection-X not found")
+
+	// Setup again, but this time use a non-existent trade connection
+	msg, _ = s.SetupTestCreateTradeRoute()
+	msg.StrideToTradeConnectionId = "connection-Y"
+
+	_, err = s.GetMsgServer().CreateTradeRoute(sdk.WrapSDKContext(s.Ctx), &msg)
+	s.Require().ErrorContains(err, "unable to register the trade ICA account: connection connection-Y not found")
+}
+
+// Tests creating a trade route where the ICA registration step fails
+func (s *KeeperTestSuite) TestCreateTradeRoute_Failure_UnableToRegisterICA() {
+	msg, expectedRoute := s.SetupTestCreateTradeRoute()
+
+	// Disable ICA middleware for the trade channel so the ICA fails
+	// TODO [DYDX]: Replace with new formatter
+	tradeAccount := expectedRoute.TradeAccount
+	tradeOwner := types.FormatICAAccountOwner(tradeAccount.ChainId, types.ICAAccountType_CONVERTER_TRADE)
```

Comment:
So this is purposefully the wrong ICAAccountOwner setup for the CONVERTER_TRADE type?

Reply:
yeah the other hadn't been merged yet